### ISSUE TITLE
Diagnostic when a trait is not in scope no longer tells you to add a use, instead tells you to box the receiver.

### ISSUE THREAD

**Issue by 5225225** [rust-lang/rust#84272]
Given the following code: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=057c1715bd0bea7866420e59e5a070f1

```rust
// Not including std::hash::Hasher, which is why the build fails
use std::hash::BuildHasher;

fn next_u64() -> u64 {
    let bh = std::collections::hash_map::RandomState::new();
    let h = bh.build_hasher();
    h.finish()
}

fn main() {}
```

The current (`nightly`) output is:

```
   Compiling playground v0.0.1 (/playground)
error[E0599]: no method named `finish` found for struct `DefaultHasher` in the current scope
   --> src/main.rs:7:7
    |
7   |     h.finish()
    |       ^^^^^^ method not found in `DefaultHasher`
    |
help: consider wrapping the receiver expression with the appropriate type
    |
7   |     Box::new(h).finish()
    |     ^^^^^^^^^ ^
help: consider wrapping the receiver expression with the appropriate type
    |
7   |     Box::new(&mut h).finish()
    |     ^^^^^^^^^^^^^  ^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0599`.
error: could not compile `playground`

To learn more, run the command again with --verbose.
```

On stable, there was a line telling you which trait you need to import in order to call the function.

```
   Compiling playground v0.0.1 (/playground)
error[E0599]: no method named `finish` found for struct `DefaultHasher` in the current scope
   --> src/main.rs:7:7
    |
7   |     h.finish()
    |       ^^^^^^ method not found in `DefaultHasher`
    |
    = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope; perhaps add a `use` for it:
    |
2   | use std::hash::Hasher;
    |

error: aborting due to previous error

For more information about this error, try `rustc --explain E0599`.
error: could not compile `playground`

To learn more, run the command again with --verbose.
```

Also, boxing the object here isn't helpful, so it shouldn't be suggested.

Bisected this, and I'm *assuming* that the relevant issue is https://github.com/rust-lang/rust/pull/83667, but I'm not sure how I would verify that.

---

searched nightlies: from nightly-2021-03-30 to nightly-2021-03-31
regressed nightly: nightly-2021-03-31
searched commits: from https://github.com/rust-lang/rust/commit/48691ea6e639640f110b43e33d4aba1f07e7415c to https://github.com/rust-lang/rust/commit/74874a690bc95443292496ff5df5cc5c8cb56e0b
regressed commit: https://github.com/rust-lang/rust/commit/16156fb2787f745e57504197bd7fe38b69c6cbea

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.0</summary>

Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --test-dir=. --start=2021-03-30 --end=2021-03-31 --preserve --script=./test.sh 
```

(where test.sh is
```sh
#!/bin/bash

cargo check 2>&1 | grep "std::hash::Hasher"
```
)
</details>

**Comment by jyn514** [comment]
cc @estebank

**Comment by estebank** [comment]
Removing these early return would likely give you the suggestion, but given the current state of the code, I'm not sure we can give the "always correct" suggestion.

**Comment by Stupremee** [comment]
Assigning `P-high` as [discussed as part of the Prioritization Working Group procedure](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2384272.20Diagnostic.20when.20a.20trait.20is.20not.20in.20scope.20no.20longer.20te.E2.80.A6/near/235001788) and removing `I-prioritize`.

**Comment by estebank** [comment]
I have a fix for this in #84499.
